odoo/addons/test_read_group/tests/test_related.py

```diff
@@ -0,0 +1,44 @@
+from odoo.tests import common
+
+
+class TestRelated(common.TransactionCase):
+    """ Test grouping by related non-stored fields. """
+
+    def test_related(self):
+        Partner = self.env['res.partner']
+        Model1 = self.env['test_read_group.related1']
+        Model2 = self.env['test_read_group.related2']
+        Model3 = self.env['test_read_group.related3']
+
+        country = self.env.ref('base.ru')
+        state = self.env['res.country.state'].create({
+            'name': 'Bashkortostan',
+            'code': '02-RUS',
+            'country_id': country.id,
+        })
+
+        for _ in range(2):
+            Model3.create({
+                'partner_id': Partner.create({
+                    'name': 'Ivan',
+                    'country_id': country.id,
+                    'state_id': state.id,
+                }).id
+            })
+
+        self.make_test(Model3, 'state_code')
+        self.make_test(Model3, 'state_code2')
+        self.make_test(Model3, 'state_stored_code')
+
+        for model in (Model2, Model3):
+            self.make_test(model, 'country_code2')
+            self.make_test(model, 'partner_city')
+
+        for model in (Model1, Model2, Model3):
+            self.make_test(model, 'country_code')
```

Comment:
Maybe add a test where `read_group()` fails because of incorrect argument (non-stored related field with `compute_sudo=False` and non-autojoin relation).